coderd/prometheusmetrics/prometheusmetrics.go

```diff
@@ -106,3 +117,159 @@ func Workspaces(ctx context.Context, registerer prometheus.Registerer, db databa
 	}()
 	return cancelFunc, nil
 }
+
+// Agents tracks the total number of workspaces with labels on status.
+func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Registerer, db database.Store, coordinator *atomic.Pointer[tailnet.Coordinator], derpMap *tailcfg.DERPMap, agentInactiveDisconnectTimeout, duration time.Duration) (context.CancelFunc, error) {
+	if duration == 0 {
+		duration = 1 * time.Minute
+	}
+
+	agentsGauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agents",
+		Name:      "up",
+		Help:      "The number of active agents per workspace.",
+	}, []string{"username", "workspace_name"})
+	err := registerer.Register(agentsGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentsConnectionsGauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agents",
+		Name:      "connections",
+		Help:      "Agent connections with statuses.",
+	}, []string{"agent_name", "username", "workspace_name", "status", "lifecycle_state", "tailnet_node"})
+	err = registerer.Register(agentsConnectionsGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentsConnectionLatenciesGauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agents",
+		Name:      "connection_latencies_seconds",
+		Help:      "Agent connection latencies in seconds.",
+	}, []string{"agent_id", "username", "workspace_name", "derp_region", "preferred"})
+	err = registerer.Register(agentsConnectionLatenciesGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentsAppsGauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agents",
+		Name:      "apps",
+		Help:      "Agent applications with statuses.",
+	}, []string{"agent_name", "username", "workspace_name", "app_name", "health"})
+	err = registerer.Register(agentsAppsGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	// nolint:gocritic // Prometheus must collect metrics for all Coder users.
+	ctx, cancelFunc := context.WithCancel(dbauthz.AsSystemRestricted(ctx))
+	ticker := time.NewTicker(duration)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			logger.Debug(ctx, "Collect agent metrics now")
+
+			workspaceRows, err := db.GetWorkspaces(ctx, database.GetWorkspacesParams{
+				AgentInactiveDisconnectTimeoutSeconds: int64(agentInactiveDisconnectTimeout.Seconds()),
+			})
+			if err != nil {
+				logger.Error(ctx, "can't get workspace rows", slog.Error(err))
+				continue
+			}
+
+			agentsGauge.Reset()
+			agentsConnectionsGauge.Reset()
+			agentsConnectionLatenciesGauge.Reset()
+			agentsAppsGauge.Reset()
```

Comment:
This feels like a race condition. We reset the gauges, then loop over something that does db calls.

The loop can be mid way through when a scrape happens, and the numbers will be incorrect.

Reply:
One cheap way to fix this is make our own `prometheus.Collector` and only update the values of the metrics at the end of the function. So we store the last values while we compute the new values, and always return the last completed dataset.

Reply:
Gauges are being reset on every loop iteration before collecting any metrics. In theory, it's possible that a scrape happens after the reset but before the end of the iteration, so metrics may disappear.

I will look into a custom collector. BTW most likely the same issue affects `prometheusmetrics.Workspaces()`.

Reply:
Yea, that one at least doesn't do db calls in the increment, but it does affect that too.
This is my first time looking at this.

Reply:
> This feels like a race condition. We reset the gauges, then loop over something that does db calls.

I implemented a wrapper to mitigate concurrency risks. It seems to be the simplest way to address it. `Collect()` will wait a bit if there is a "flushing" procedure ongoing.